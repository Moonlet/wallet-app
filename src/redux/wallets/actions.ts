import { HDWallet } from '../../core/wallet/hd-wallet/hd-wallet';
import { Blockchain } from '../../core/blockchain/types';
import { WalletType } from '../../core/wallet/types';
import { IWalletState, IAccountState } from './state';
import { IAction } from '../types';
import { Dispatch } from 'react';
import { IReduxState } from '../state';
import { getChainId } from '../app/selectors';
import { appSwitchWallet, openBottomSheet, closeBottomSheet } from '../app/actions';
import uuidv4 from 'uuid/v4';
import { storeEncrypted, deleteFromStorage } from '../../core/secure/storage';
import { getBlockchain } from '../../core/blockchain/blockchain-factory';
import { WalletFactory } from '../../core/wallet/wallet-factory';
import { selectCurrentWallet } from './selectors';
import { HWVendor, HWModel, HWConnection } from '../../core/wallet/hw-wallet/types';
import {
    verifyAddressOnDevice,
    featureNotSupported,
    toInitialState,
    connectInProgress
} from '../screens/connectHardwareWallet/actions';
import { HWWalletFactory } from '../../core/wallet/hw-wallet/hw-wallet-factory';
import { NavigationScreenProp, NavigationState, NavigationActions } from 'react-navigation';
import { LedgerWallet } from '../../core/wallet/hw-wallet/ledger/ledger-wallet';
import { translate } from '../../core/i18n';
import { REVIEW_TRANSACTION } from '../screens/send/actions';
import { BottomSheetType, ICurrentAccount } from '../app/state';
import { REHYDRATE } from 'redux-persist';
import { TokenType, ITokenConfig } from '../../core/blockchain/types/token';
import BigNumber from 'bignumber.js';

// actions consts
export const WALLET_ADD = 'WALLET_ADD';
export const WALLET_DELETE = 'WALLET_DELETE';
export const WALLET_CHANGE_NAME = 'WALLET_CHANGE_NAME';
export const ACCOUNT_GET_BALANCE = 'ACCOUNT_GET_BALANCE';
export const TRANSACTION_PUBLISHED = 'TRANSACTION_PUBLISHED';
export const ACCOUNT_ADD = 'ACCOUNT_ADD';
export const ACCOUNT_REMOVE = 'ACCOUNT_REMOVE';
export const TOGGLE_TOKEN_ACTIVE = 'TOGGLE_TOKEN_ACTIVE';
export const UPDATE_TOKEN_ORDER = 'UPDATE_TOKEN_ORDER';
export const REMOVE_TOKEN = 'REMOVE_TOKEN';
export const ADD_TOKEN = 'ADD_TOKEN';
export const WALLET_SELECT_ACCOUNT = 'WALLET_SELECT_ACCOUNT';

// action creators
export const addWallet = (walletData: IWalletState) => {
    return {
        type: WALLET_ADD,
        data: walletData
    };
};

export const switchSelectedAccount = (currentAccount: ICurrentAccount) => (
    dispatch: Dispatch<IAction<any>>,
    getState: () => IReduxState
) => {
    const wallet = selectCurrentWallet(getState());
    if (wallet === undefined) {
        return;
    }
    dispatch({
        type: WALLET_SELECT_ACCOUNT,
        data: {
            walletId: wallet.id,
            currentAccount
        }
    });
};

export const updateReduxState = (state: IReduxState) => dispatch => {
    dispatch({
        type: REHYDRATE,
        payload: state
    });
};

export const addAccount = (walletId: string, blockchain: Blockchain, account: IAccountState) => {
    return {
        type: ACCOUNT_ADD,
        data: { walletId, account, blockchain }
    };
};

export const removeAccount = (walletId: string, blockchain: Blockchain, account: IAccountState) => {
    return {
        type: ACCOUNT_REMOVE,
        data: { walletId, account, blockchain }
    };
};

export const createHWWallet = (
    deviceId: string,
    deviceVendor: HWVendor,
    deviceModel: HWModel,
    connectionType: HWConnection,
    blockchain: Blockchain,
    navigation: NavigationScreenProp<NavigationState>
) => async (dispatch: Dispatch<IAction<any>>, getState: () => IReduxState) => {
    try {
        const walletId: string = uuidv4();

        // in case you replace your connected ledger reset message
        dispatch(toInitialState());
        dispatch(connectInProgress());

        const wallet = await HWWalletFactory.get(
            deviceVendor,
            deviceModel,
            deviceId,
            connectionType
        );

        await (wallet as LedgerWallet).onAppOpened(blockchain);

        dispatch(verifyAddressOnDevice(true));
        const accounts: IAccountState[] = await wallet.getAccounts(blockchain, 0);
        accounts[0].selected = true;
        const walletData: IWalletState = {
            id: walletId,
            hwOptions: {
                deviceId,
                deviceVendor,
                deviceModel,
                connectionType
            },
            name: `Wallet ${Object.keys(getState().wallets).length + 1}`,
            type: WalletType.HW,
            accounts
        };

        dispatch(addWallet(walletData));

        dispatch(appSwitchWallet(walletId));
        navigation.navigate(
            'MainNavigation',
            {},
            NavigationActions.navigate({ routeName: 'Dashboard' })
        );
        dispatch(toInitialState());
    } catch (e) {
        // this might not be the best place
        if (e === translate('CreateHardwareWallet.notSupported')) {
            dispatch(featureNotSupported());
        }
        throw new Error(e);
    }
};

export const createHDWallet = (mnemonic: string, password: string, callback?: () => any) => async (
    dispatch: Dispatch<IAction<any>>,
    getState: () => IReduxState
) => {
    try {
        const wallet = new HDWallet(mnemonic);

        // generate initial accounts for each blockchain
        Promise.all([
            wallet.getAccounts(Blockchain.ETHEREUM, 0),
            wallet.getAccounts(Blockchain.ETHEREUM, 1),
            wallet.getAccounts(Blockchain.ZILLIQA, 0),
            wallet.getAccounts(Blockchain.ZILLIQA, 1)
        ]).then(async data => {
            data[0][0].selected = true;
            //    data[2][0].selected = true;
            const walletId = uuidv4();
            const accounts: IAccountState[] = data.reduce((out, acc) => out.concat(acc), []);

            dispatch(
                addWallet({
                    id: walletId,
                    name: `Wallet ${Object.keys(getState().wallets).length + 1}`,
                    type: WalletType.HD,
                    accounts
                })
            );

            await storeEncrypted(mnemonic, walletId, password);

            dispatch(appSwitchWallet(walletId));
            callback && callback();
        });
    } catch (e) {
        // console.log(e);
        // TODO best way to handle this?
    }
    // TODO  - error handling
};

// will check balance for a coin or all coins if needed
export const getBalance = (
    blockchain: Blockchain,
    address: string,
    token: string = undefined,
    force: boolean = false
) => async (dispatch, getState: () => IReduxState) => {
    const state = getState();
    const wallet = selectCurrentWallet(state);
    if (wallet === undefined) {
        return;
    }
    const account = wallet.accounts.filter(
        acc => acc.address === address && acc.blockchain === blockchain
    )[0];

    // console.log('getBalance', { blockchain, address, token });
    if (token) {
        const balanceInProgress = account?.tokens[token]?.balance?.inProgress;
        const balanceTimestamp = account?.tokens[token]?.balance?.timestamp || 0;

        if (force || (!balanceInProgress && balanceTimestamp + 10 * 3600 < Date.now())) {
            const data = {
                walletId: wallet.id,
                address,
                token,
                blockchain
            };

            dispatch({
                type: ACCOUNT_GET_BALANCE,
                data,
                inProgress: true
            });
            try {
                const chainId = getChainId(state, account.blockchain);
                const tokenInfo = account.tokens[token];

                let balance;
                switch (tokenInfo.type) {
                    case TokenType.NATIVE:
                        balance = await getBlockchain(blockchain)
                            .getClient(chainId)
                            .getBalance(address);
                        break;
                    default:
                        const client = getBlockchain(blockchain).getClient(chainId);
                        if (client.tokens[tokenInfo.type]) {
                            balance = await client.tokens[tokenInfo.type].getBalance(
                                tokenInfo.contractAddress,
                                address
                            );
                        } else {
                            throw new Error(
                                `Token Type (${tokenInfo.type}) not handled for blockchain ${blockchain}.`
                            );
                        }
                }

                dispatch({
                    type: ACCOUNT_GET_BALANCE,
                    data: {
                        ...data,
                        balance
                    }
                });
            } catch (error) {
                dispatch({
                    type: ACCOUNT_GET_BALANCE,
                    data,
                    error
                });
            }
        }
    } else {
        // call get balance for all tokens
        Object.keys(account.tokens || {}).map(tokenSymbol => {
            // console.log(`getBalance(${blockchain}, ${address}, ${tokenSymbol}, ${force})`);
            getBalance(blockchain, address, tokenSymbol, force)(dispatch, getState);
        });
    }
};

export const sendTransferTransaction = (
    account: IAccountState,
    toAddress: string,
    amount: string,
    token: string,
    feeOptions: any,
    password: string,
    navigation: NavigationScreenProp<NavigationState>,
    goBack: boolean = true
) => async (dispatch, getState: () => IReduxState) => {
    const state = getState();
    const chainId = getChainId(state, account.blockchain);

    const appWallet = selectCurrentWallet(state);

    try {
        const wallet = await WalletFactory.get(appWallet.id, appWallet.type, {
            pass: password,
            deviceVendor: appWallet.hwOptions?.deviceVendor,
            deviceModel: appWallet.hwOptions?.deviceModel,
            deviceId: appWallet.hwOptions?.deviceId,
            connectionType: appWallet.hwOptions?.connectionType
        }); // encrypted string: pass)
        const blockchainInstance = getBlockchain(account.blockchain);

        const nonce = await blockchainInstance.getClient(chainId).getNonce(account.address);

        const tx = blockchainInstance.transaction.buildTransferTransaction({
            chainId,
            account,
            toAddress,
            amount: blockchainInstance.account.amountToStd(amount, account.tokens[token].decimals),
            token,
            nonce,
            gasPrice: new BigNumber(feeOptions.gasPrice),
            gasLimit: new BigNumber(feeOptions.gasLimit).toNumber()
        });

        if (appWallet.type === WalletType.HW) {
            dispatch(
                openBottomSheet(BottomSheetType.LEDGER_SIGN_MESSAGES, {
                    blockchain: account.blockchain
                })
            );

            await (wallet as LedgerWallet).onAppOpened(account.blockchain);

            dispatch({
                type: REVIEW_TRANSACTION,
                data: true
            });
        }
        const transaction = await wallet.sign(account.blockchain, account.index, tx);

        const publish = await getBlockchain(account.blockchain)
            .getClient(chainId)
            .sendTransaction(transaction);
        if (publish) {
            dispatch({
                type: TRANSACTION_PUBLISHED,
                data: {
                    hash: transaction,
                    tx,
                    walletId: appWallet.id
                }
            });
            if (appWallet.type === WalletType.HW) {
                dispatch({
                    type: REVIEW_TRANSACTION,
                    data: false
                });
                dispatch(closeBottomSheet());
            }
            goBack && navigation.goBack();
            return;
        }
    } catch (e) {
        throw new Error(e);
    }
};

export const deleteWallet = (walletId: string) => (
    dispatch: Dispatch<IAction<any>>,
    getState: () => IReduxState
) => {
    const state = getState();
    if (state.app.currentWalletId === walletId) {
        const nextWallet = Object.values(state.wallets).find(wallet => wallet.id !== walletId);
        const nextWalletId = nextWallet ? nextWallet.id : '';
        dispatch(appSwitchWallet(nextWalletId));
    }
    dispatch({
        type: WALLET_DELETE,
        data: walletId
    });
    deleteFromStorage(walletId);
};

export const updateWalletName = (walletId: string, newName: string) => {
    return {
        type: WALLET_CHANGE_NAME,
        data: { walletId, newName }
    };
};

export const toggleTokenActive = (
    walletId: string,
    account: IAccountState,
    token: ITokenConfig
) => {
    return {
        type: TOGGLE_TOKEN_ACTIVE,
        data: { walletId, account, token }
    };
};

export const updateTokenOrder = (
    walletId: string,
    account: IAccountState,
    tokens: ITokenConfig[]
) => {
    return {
        type: UPDATE_TOKEN_ORDER,
        data: { walletId, account, tokens }
    };
};

export const removeToken = (walletId: string, account: IAccountState, token: ITokenConfig) => {
    return {
        type: REMOVE_TOKEN,
        data: { walletId, account, token }
    };
};

export const addToken = (walletId: string, account: IAccountState, token: ITokenConfig) => {
    return {
        type: ADD_TOKEN,
        data: { walletId, account, token }
    };
};
