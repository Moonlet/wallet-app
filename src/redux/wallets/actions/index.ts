export * from './pos-actions';
export * from './wallet-actions';
export * from './pos-actions/solana';
