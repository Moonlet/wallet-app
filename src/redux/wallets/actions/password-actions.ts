import { IWalletState } from '../state';
import { Dispatch } from 'react';
import { IReduxState } from '../../state';
import { storeEncrypted, readEncrypted } from '../../../core/secure/storage/storage';

import { getEncryptionKey, generateEncryptionKey } from '../../../core/secure/keychain/keychain';

export const changePIN = (newPassword: string, oldPassword: string) => async (
    dispatch: Dispatch<any>,
    getState: () => IReduxState
) => {
    const state = getState();

    const encryptionKey = await getEncryptionKey(oldPassword);
    const newEncryptionKey = await generateEncryptionKey(newPassword);

    Object.values(state.wallets).map(async (wallet: IWalletState) => {
        const walletId = wallet.id;

        const mnemonic = await readEncrypted(walletId, encryptionKey);

        await storeEncrypted(mnemonic, walletId, newEncryptionKey);
    });
};
