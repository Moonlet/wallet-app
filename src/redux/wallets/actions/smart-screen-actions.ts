import { IAccountState } from '../state';
import { IFeeOptions, ITransactionExtraFields } from '../../../core/blockchain/types';
import { Dispatch } from 'react';
import { IAction } from '../../types';
import { IReduxState } from '../../state';
import { getChainId } from '../../preferences/selectors';
import { getBlockchain } from '../../../core/blockchain/blockchain-factory';
import { IValidator } from '../../../core/blockchain/types/stats';
import {
    openProcessTransactions,
    setProcessTransactions
} from '../../ui/process-transactions/actions';
import { TransactionStatus } from '../../../core/wallet/types';
import { cloneDeep } from 'lodash';
import { captureException as SentryCaptureException } from '@sentry/react-native';
import { PosBasicActionType } from '../../../core/blockchain/types/token';

export const swap = (
    account: IAccountState,
    validators: {
        validator: IValidator;
        amount: string;
    }[],
    token: string,
    feeOptions: IFeeOptions,
    extraFields: ITransactionExtraFields
) => async (dispatch: Dispatch<IAction<any>>, getState: () => IReduxState) => {
    const state = getState();
    const chainId = getChainId(state, account.blockchain);

    // TODO CHANGE PROPERLY to SWAP
    try {
        const extra: ITransactionExtraFields = {
            ...extraFields
        };
        const blockchainInstance = getBlockchain(account.blockchain);

        dispatch(openProcessTransactions());

        const txs = await blockchainInstance.transaction.buildPosTransaction(
            {
                chainId,
                account,
                validators: validators as any,
                amount: '0',
                token,
                feeOptions:
                    feeOptions?.gasPrice && feeOptions?.gasLimit
                        ? {
                              gasPrice: feeOptions.gasPrice.toString(),
                              gasLimit: feeOptions.gasLimit.toString()
                          }
                        : undefined,
                extraFields: extra
            },
            PosBasicActionType.UNSTAKE
        );
        dispatch(
            setProcessTransactions(
                cloneDeep(txs).map(tx => {
                    tx.status = TransactionStatus.CREATED;
                    return tx;
                })
            )
        );
    } catch (errorMessage) {
        SentryCaptureException(new Error(JSON.stringify(errorMessage)));
    }
};
