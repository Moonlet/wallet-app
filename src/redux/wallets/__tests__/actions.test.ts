import { addWallet, createHDWallet } from '../actions';

describe('wallet actions', () => {
    test('addWallet actions return correct object', () => {
        // @ts-ignore
        expect(addWallet({ data: 'test' })).toEqual({ data: { data: 'test' }, type: 'WALLET_ADD' });
    });

    test('createHDWallet working properly', async () => {
        const dispatch = jest.fn();
        const getState = jest.fn((): any => ({ wallets: [] }));
        const callback = jest.fn();

        const mnemonic =
            'panic club above clarify orbit resist illegal feel bus remember aspect field test bubble dog trap awesome hand room rice heavy idle faint salmon';

        const response = {
            data: {
                accounts: [
                    {
                        address: '0x151250cB2bB034247c8ffBeD91B7Fb1e9f3e9FFE',
                        blockchain: 'ETHEREUM',
                        index: 0,
                        publicKey:
                            '22d185a1eda6240e2004a4794e50daff53eb3e5194bf21d4cca3be002688566ee2af4d5724dbe507646649193e5f8368ee63fb257147fbb2bafed87e63717c68'
                    },
                    {
                        address: 'zil1kqj0fp7ja409llrjt4q4tja4uvue8vtp65gahs',
                        blockchain: 'ZILLIQA',
                        index: 0,
                        publicKey:
                            '0310a8a9ad39c2c9a878bf1a42895ee122806d1f4cf510f604d274b04f5a5854cc'
                    }
                ],
                id: 'hdWallet',
                type: 'HD'
            },
            type: 'WALLET_ADD'
        };

        await createHDWallet(mnemonic, callback)(dispatch, getState);

        expect(callback).toHaveBeenCalled();
        expect(dispatch).toBeCalledWith(response);
    });
});
