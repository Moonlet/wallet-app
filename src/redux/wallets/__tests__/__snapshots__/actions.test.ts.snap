// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`wallet actions addWallet actions return correct object 1`] = `
Object {
  "data": "1",
  "type": "SELECT_WALLET",
}
`;

exports[`wallet actions createHDWallet working properly 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "type": "OPEN_LOADING_MODAL",
      },
    ],
    Array [
      Object {
        "data": Object {
          "accounts": Array [
            Object {
              "address": "zil1kqj0fp7ja409llrjt4q4tja4uvue8vtp65gahs",
              "blockchain": "ZILLIQA",
              "index": 0,
              "publicKey": "0310a8a9ad39c2c9a878bf1a42895ee122806d1f4cf510f604d274b04f5a5854cc",
              "selected": true,
              "tokens": Object {
                "ZIL": Object {
                  "active": true,
                  "decimals": 12,
                  "logo": Object {
                    "uri": "https://raw.githubusercontent.com/atomiclabs/cryptocurrency-icons/master/128/icon/zil.png",
                  },
                  "name": "Zilliqa",
                  "order": 0,
                  "symbol": "ZIL",
                  "type": "NATIVE",
                  "ui": Object {
                    "decimals": 3,
                    "tokenScreenComponent": "DEFAULT",
                  },
                  "units": Object {
                    "LI": "1000000",
                    "QA": "1",
                    "ZIL": "1000000000000",
                  },
                },
              },
            },
            Object {
              "address": "zil1lt5t8hkzgmzfwqp2s234qpr3hn9un9zdck56wm",
              "blockchain": "ZILLIQA",
              "index": 1,
              "publicKey": "0362f31a424e62c75b38534ce16a1b607dd216eae22aa4add9042b16d6750e43e8",
              "selected": false,
              "tokens": Object {
                "ZIL": Object {
                  "active": true,
                  "decimals": 12,
                  "logo": Object {
                    "uri": "https://raw.githubusercontent.com/atomiclabs/cryptocurrency-icons/master/128/icon/zil.png",
                  },
                  "name": "Zilliqa",
                  "order": 0,
                  "symbol": "ZIL",
                  "type": "NATIVE",
                  "ui": Object {
                    "decimals": 3,
                    "tokenScreenComponent": "DEFAULT",
                  },
                  "units": Object {
                    "LI": "1000000",
                    "QA": "1",
                    "ZIL": "1000000000000",
                  },
                },
              },
            },
            Object {
              "address": "zil1sdlr86lc6q53q6yqz03eq8chq08jdfxxva2ja9",
              "blockchain": "ZILLIQA",
              "index": 2,
              "publicKey": "03ac450f102e107fb033dfb925b75b63f6855a7933b9aec26742971bdb64939fb3",
              "selected": false,
              "tokens": Object {
                "ZIL": Object {
                  "active": true,
                  "decimals": 12,
                  "logo": Object {
                    "uri": "https://raw.githubusercontent.com/atomiclabs/cryptocurrency-icons/master/128/icon/zil.png",
                  },
                  "name": "Zilliqa",
                  "order": 0,
                  "symbol": "ZIL",
                  "type": "NATIVE",
                  "ui": Object {
                    "decimals": 3,
                    "tokenScreenComponent": "DEFAULT",
                  },
                  "units": Object {
                    "LI": "1000000",
                    "QA": "1",
                    "ZIL": "1000000000000",
                  },
                },
              },
            },
            Object {
              "address": "zil1pdgzy7mne6qmv6z298u2w4fpd7272yuuh5vctc",
              "blockchain": "ZILLIQA",
              "index": 3,
              "publicKey": "020acf03d556082f341b650ec632592809638a6e2e7886e9d65e4164ed6fa3bb8a",
              "selected": false,
              "tokens": Object {
                "ZIL": Object {
                  "active": true,
                  "decimals": 12,
                  "logo": Object {
                    "uri": "https://raw.githubusercontent.com/atomiclabs/cryptocurrency-icons/master/128/icon/zil.png",
                  },
                  "name": "Zilliqa",
                  "order": 0,
                  "symbol": "ZIL",
                  "type": "NATIVE",
                  "ui": Object {
                    "decimals": 3,
                    "tokenScreenComponent": "DEFAULT",
                  },
                  "units": Object {
                    "LI": "1000000",
                    "QA": "1",
                    "ZIL": "1000000000000",
                  },
                },
              },
            },
            Object {
              "address": "zil1kpqpw78el6zqnmmkvxtcgzh7jyjmktfwtr7nwg",
              "blockchain": "ZILLIQA",
              "index": 4,
              "publicKey": "0375d579fddf8e8088b8d195844b478f8d4b3cb7004b76c5c2de2ed2f47a439de5",
              "selected": false,
              "tokens": Object {
                "ZIL": Object {
                  "active": true,
                  "decimals": 12,
                  "logo": Object {
                    "uri": "https://raw.githubusercontent.com/atomiclabs/cryptocurrency-icons/master/128/icon/zil.png",
                  },
                  "name": "Zilliqa",
                  "order": 0,
                  "symbol": "ZIL",
                  "type": "NATIVE",
                  "ui": Object {
                    "decimals": 3,
                    "tokenScreenComponent": "DEFAULT",
                  },
                  "units": Object {
                    "LI": "1000000",
                    "QA": "1",
                    "ZIL": "1000000000000",
                  },
                },
              },
            },
            Object {
              "address": "0x151250cB2bB034247c8ffBeD91B7Fb1e9f3e9FFE",
              "blockchain": "ETHEREUM",
              "index": 0,
              "publicKey": "22d185a1eda6240e2004a4794e50daff53eb3e5194bf21d4cca3be002688566ee2af4d5724dbe507646649193e5f8368ee63fb257147fbb2bafed87e63717c68",
              "selected": true,
              "tokens": Object {
                "ETH": Object {
                  "active": true,
                  "decimals": 18,
                  "logo": Object {
                    "uri": "https://raw.githubusercontent.com/atomiclabs/cryptocurrency-icons/master/128/icon/eth.png",
                  },
                  "name": "Ethereum",
                  "order": 0,
                  "symbol": "ETH",
                  "type": "NATIVE",
                  "ui": Object {
                    "decimals": 4,
                    "tokenScreenComponent": "DEFAULT",
                  },
                  "units": Object {
                    "ETH": "1000000000000000000",
                    "GWEI": "1000000000",
                    "WEI": "1",
                  },
                },
              },
            },
            Object {
              "address": "0x6D179861ea6C879269Fa65b71119e86b3e075c64",
              "blockchain": "ETHEREUM",
              "index": 1,
              "publicKey": "d95dadc866a4acbbf37fb7347f26ddf59955c0b978f7356861e2e7e7a17eaaa07e632630ee164969b84a196e724acbd0ad9b230276c694c0de2fb9fc05239398",
              "selected": false,
              "tokens": Object {
                "ETH": Object {
                  "active": true,
                  "decimals": 18,
                  "logo": Object {
                    "uri": "https://raw.githubusercontent.com/atomiclabs/cryptocurrency-icons/master/128/icon/eth.png",
                  },
                  "name": "Ethereum",
                  "order": 0,
                  "symbol": "ETH",
                  "type": "NATIVE",
                  "ui": Object {
                    "decimals": 4,
                    "tokenScreenComponent": "DEFAULT",
                  },
                  "units": Object {
                    "ETH": "1000000000000000000",
                    "GWEI": "1000000000",
                    "WEI": "1",
                  },
                },
              },
            },
            Object {
              "address": "0x99B494729398F27acF32a3237eC87c1547ce33f3",
              "blockchain": "ETHEREUM",
              "index": 2,
              "publicKey": "80d9f6726b66996bef42495c8623ea7ef8c9429da5eec12c666998ed2fd9cbde86da22b8252e9b8e3e638bd7405bea4db11d4efc305c616ce7e9cececf2bd03e",
              "selected": false,
              "tokens": Object {
                "ETH": Object {
                  "active": true,
                  "decimals": 18,
                  "logo": Object {
                    "uri": "https://raw.githubusercontent.com/atomiclabs/cryptocurrency-icons/master/128/icon/eth.png",
                  },
                  "name": "Ethereum",
                  "order": 0,
                  "symbol": "ETH",
                  "type": "NATIVE",
                  "ui": Object {
                    "decimals": 4,
                    "tokenScreenComponent": "DEFAULT",
                  },
                  "units": Object {
                    "ETH": "1000000000000000000",
                    "GWEI": "1000000000",
                    "WEI": "1",
                  },
                },
              },
            },
            Object {
              "address": "0x0dCdBDd1aeb52d6Dc1079e8D31F24F8c9109829C",
              "blockchain": "ETHEREUM",
              "index": 3,
              "publicKey": "4707217d4ecd5bf90bd9762b3997e014155abe724967ca5162d9ed6e37b4bfa583f576f4ff371014cf64ede7993b3bc0deed630fe12d245be3f4255cb4f671ed",
              "selected": false,
              "tokens": Object {
                "ETH": Object {
                  "active": true,
                  "decimals": 18,
                  "logo": Object {
                    "uri": "https://raw.githubusercontent.com/atomiclabs/cryptocurrency-icons/master/128/icon/eth.png",
                  },
                  "name": "Ethereum",
                  "order": 0,
                  "symbol": "ETH",
                  "type": "NATIVE",
                  "ui": Object {
                    "decimals": 4,
                    "tokenScreenComponent": "DEFAULT",
                  },
                  "units": Object {
                    "ETH": "1000000000000000000",
                    "GWEI": "1000000000",
                    "WEI": "1",
                  },
                },
              },
            },
            Object {
              "address": "0x7bEE31e0Da267B1874ADAD851c6b925A11104cAA",
              "blockchain": "ETHEREUM",
              "index": 4,
              "publicKey": "4a8ec646a13b1b2d7a49e6f578191a81da860c8461077110b3bd05fb10025ac9caecbbbeb96819ee4114cc1aca3ab633b9fd15e21d256f8fa0f0f05f4bc2f802",
              "selected": false,
              "tokens": Object {
                "ETH": Object {
                  "active": true,
                  "decimals": 18,
                  "logo": Object {
                    "uri": "https://raw.githubusercontent.com/atomiclabs/cryptocurrency-icons/master/128/icon/eth.png",
                  },
                  "name": "Ethereum",
                  "order": 0,
                  "symbol": "ETH",
                  "type": "NATIVE",
                  "ui": Object {
                    "decimals": 4,
                    "tokenScreenComponent": "DEFAULT",
                  },
                  "units": Object {
                    "ETH": "1000000000000000000",
                    "GWEI": "1000000000",
                    "WEI": "1",
                  },
                },
              },
            },
            Object {
              "address": "cosmos1mrf865hmk3prnkhtpqt8zdwu7kpmpd6l30ws3q",
              "blockchain": "COSMOS",
              "index": 0,
              "publicKey": "02d7144d5afafde20f868a669d74920bac76c7de2750352c2b0ef8162fe8476c1d",
              "selected": false,
              "tokens": Object {
                "ATOM": Object {
                  "active": true,
                  "decimals": 6,
                  "logo": Object {
                    "uri": "https://raw.githubusercontent.com/atomiclabs/cryptocurrency-icons/master/128/icon/atom.png",
                  },
                  "name": "Atom",
                  "order": 0,
                  "symbol": "ATOM",
                  "symbolMap": Object {
                    "gaia-13007": "umuon",
                  },
                  "type": "NATIVE",
                  "ui": Object {
                    "decimals": 3,
                    "tokenScreenComponent": "DELEGATE",
                  },
                  "units": Object {
                    "ATOM": "1000000",
                    "UATOM": "1",
                  },
                },
              },
            },
            Object {
              "address": "cosmos1fptf3l633k4kplhslqd7fksfke5kz2saxzupl8",
              "blockchain": "COSMOS",
              "index": 1,
              "publicKey": "03c47735b7c423d6ad9c9be4592f2879c6c30a7357186a0d7e71735abf3bbf09f8",
              "selected": false,
              "tokens": Object {
                "ATOM": Object {
                  "active": true,
                  "decimals": 6,
                  "logo": Object {
                    "uri": "https://raw.githubusercontent.com/atomiclabs/cryptocurrency-icons/master/128/icon/atom.png",
                  },
                  "name": "Atom",
                  "order": 0,
                  "symbol": "ATOM",
                  "symbolMap": Object {
                    "gaia-13007": "umuon",
                  },
                  "type": "NATIVE",
                  "ui": Object {
                    "decimals": 3,
                    "tokenScreenComponent": "DELEGATE",
                  },
                  "units": Object {
                    "ATOM": "1000000",
                    "UATOM": "1",
                  },
                },
              },
            },
            Object {
              "address": "cosmos1t257fk07eevpr0kr8rld8dyp0a2e9hmh9tqaad",
              "blockchain": "COSMOS",
              "index": 2,
              "publicKey": "03b8edd9cbacc9092a51283d8ff629fd7732bd8a3e3a808c7c35628c667896154e",
              "selected": false,
              "tokens": Object {
                "ATOM": Object {
                  "active": true,
                  "decimals": 6,
                  "logo": Object {
                    "uri": "https://raw.githubusercontent.com/atomiclabs/cryptocurrency-icons/master/128/icon/atom.png",
                  },
                  "name": "Atom",
                  "order": 0,
                  "symbol": "ATOM",
                  "symbolMap": Object {
                    "gaia-13007": "umuon",
                  },
                  "type": "NATIVE",
                  "ui": Object {
                    "decimals": 3,
                    "tokenScreenComponent": "DELEGATE",
                  },
                  "units": Object {
                    "ATOM": "1000000",
                    "UATOM": "1",
                  },
                },
              },
            },
            Object {
              "address": "cosmos14rnhtdvxgccrm54h84t7hk6md0pkk2f7mmlpw3",
              "blockchain": "COSMOS",
              "index": 3,
              "publicKey": "03c059770e2666cee4198ab56328e04db9e0299d2468cd68924b317b91dfe269df",
              "selected": false,
              "tokens": Object {
                "ATOM": Object {
                  "active": true,
                  "decimals": 6,
                  "logo": Object {
                    "uri": "https://raw.githubusercontent.com/atomiclabs/cryptocurrency-icons/master/128/icon/atom.png",
                  },
                  "name": "Atom",
                  "order": 0,
                  "symbol": "ATOM",
                  "symbolMap": Object {
                    "gaia-13007": "umuon",
                  },
                  "type": "NATIVE",
                  "ui": Object {
                    "decimals": 3,
                    "tokenScreenComponent": "DELEGATE",
                  },
                  "units": Object {
                    "ATOM": "1000000",
                    "UATOM": "1",
                  },
                },
              },
            },
            Object {
              "address": "cosmos158rxkf0wlhjfwutk63zlc68v57hzst2n4z7pe6",
              "blockchain": "COSMOS",
              "index": 4,
              "publicKey": "0245ab42ba4fdfc9d5102b94a84ad78c1480f0d1fb0561d41809c1890d48ebbda5",
              "selected": false,
              "tokens": Object {
                "ATOM": Object {
                  "active": true,
                  "decimals": 6,
                  "logo": Object {
                    "uri": "https://raw.githubusercontent.com/atomiclabs/cryptocurrency-icons/master/128/icon/atom.png",
                  },
                  "name": "Atom",
                  "order": 0,
                  "symbol": "ATOM",
                  "symbolMap": Object {
                    "gaia-13007": "umuon",
                  },
                  "type": "NATIVE",
                  "ui": Object {
                    "decimals": 3,
                    "tokenScreenComponent": "DELEGATE",
                  },
                  "units": Object {
                    "ATOM": "1000000",
                    "UATOM": "1",
                  },
                },
              },
            },
          ],
          "id": "uuid",
          "name": "Wallet 1",
          "selected": false,
          "selectedBlockchain": "ZILLIQA",
          "type": "HD",
        },
        "type": "WALLET_ADD",
      },
    ],
    Array [
      Object {
        "data": "uuid",
        "type": "SELECT_WALLET",
      },
    ],
    Array [
      Object {
        "type": "CLOSE_LOADING_MODAL",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
